Katello/katello
# Fixes #36776 - Fixed issue where composite content views promoted to …

…lifecycle environments referencing the env name would error on promotion.

Changed env replacement source.

#### What are the changes introduced in this pull request?
Composite content views containing archived repositories have their environment set to nil. When promoting this CVV to a lifecycle environment with a registry name containing the lifecycle environment's name, this would previously cause the action to error out. After the changes, if the env is unset, Katello will grab a valid environment from the first non-archived repository in the repository's content view.

#### Considerations taken when implementing this change?
I'm not 100% positive on this solution. This line is sketchy, but I haven't been able to break it in my testing:
```
environment ||= repository.content_view.repositories.first.environment
```
Is there a scenario where the archived repo's content view's first repo would belong to a different environment than the unarchived repo? Is there a better way to link an archived repo to it's unarchived version?

#### What are the testing steps for this pull request?

1. Create a product:
```
hammer product create --description "Product Busybox Container" --name Product_Busybox_Container --label "Product_Busybox_Container" --organization myorg
```
2. Create a repository and sync it:
```
hammer repository create \
  --content-type "docker" \
  --organization myorg \
  --product "Product_Busybox_Container" \
  --url "https://registry-1.docker.io/" \
  --name "library/busybox" \
  --docker-upstream-name "library/busybox" \
  --label "library_busybox" \
  --include-tags "stable, 1.35.0"
hammer repository synchronize --name "library/busybox" --product "Product_Busybox_Container" --organization myorg
```
3. Create a content view with this repo and publish it:
```
hammer content-view create --label ContentView_Busybox --name ContentView_Busybox --organization-label myorg --repository-ids <repo_id>
hammer content-view publish --organization myorg --name ContentView_Busybox
```
4. Using the UI, go to Content -> Lifecycle -> Content Views. Create a content view of type composite content view. I left auto-publish unchecked.
5. Once created, add the ContentView_Busybox CV to this composite CV.
6. Go to Content -> Lifecycle -> Lifecycle Environments. Create an environment path. Click on the newly-created environment path. On the details tab, modify the registry name pattern to `<%= repository.name %>`.
7. Create another environment path. Set its registry name pattern to `<%= lifecycle_environment.label %>/<%= repository.name %>`
8. Go to Content -> Lifecycle -> Content Views. Click on the kebab menu next to your composite content view and select promote. Select the first lifecycle environment and publish. There should be no errors because the environment isn't being referenced.
9. Refresh the page. Click on the kebab menu next to your composite content view and select promote. Select the second lifecycle environment and publish. This PR should not error out, since the environment is now being pulled from the un-archived repo instead of using the archived repo's nil environment.
